core/src/main/scala/kafka/server/ApiVersionManager.scala

```diff
@@ -112,4 +157,8 @@ class DefaultApiVersionManager(
       zkMigrationEnabled
     )
   }
+
+  override def apiVersionResponse(throttleTimeMs: Int, finalizedFeatures: Map[String, java.lang.Short], finalizedFeatureEpoch: Long): ApiVersionsResponse = {
+    throw new UnsupportedOperationException("This method is not supported in DefaultApiVersionManager, use apiVersionResponse(throttleTimeMs) instead")
```

Comment:
What's the thinking around this? It's extremely brittle to design interfaces like this.

Reply:
The `features` in `ApiVersionResponse` is retrieved directly from `MetadataCache` in BrokerServer/KafkaServer, but in ControllerServer it can only be got asynchronized and can't be unified, so I added 2 different methods here.
 I'm still checking whether there are better way to handle this, one way is to add a synchronized method to controller, another is to make `ApiVersionManager.apiVersionResponse` asynchronized, both will introduce new problems, the root cause is that `ApiVersionRequest` is treated static but finalized features is changing dynamically.

Reply:
I think @cmccabe submitted a PR to fix this here https://github.com/apache/kafka/pull/13826